### PR TITLE
Widen build_managed_expression_form to accept an ExpressionReference

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-1284

## 📝 Description
For the final set of preparatory refactors, we now update the managed expression building/editing forms to take an arbitrary `ExpressionReference` instead of a specific Question.

These flows still only allow you to pick Questions from the data referencing flow, but after this is merged we should be able to build that feature work out and have everything slot together nicely - if our plan was correct.

### Review Readiness
- [x] PR title and description provide sufficient context for reviewers
- [x] Commits are logical, self-contained, and have clear descriptions

### Testing
- [x] I have tested this change and it meets the acceptance criteria for the ticket
- [x] I need the reviewer(s) to pull and run this change locally
- [-] New (non-developer) functionality has appropriate unit and integration tests
- [x] End-to-end tests have been updated (if applicable)
- [x] Edge cases and error conditions are tested

### CODE REVIEW

File: app/common/expressions/references.py [hunk 1]
```diff
@@ -151,6 +156,30 @@ def data_type(self) -> QuestionDataType:
...
+    def data_options(self) -> QuestionDataOptions:
+        if question := self.question:
+            return question.data_options
+
+        if column := self.data_source_column:
+            return column.data_options
+
+        raise ValueError(
+            f"Can't resolve ExpressionReference {self!r} to a specific data options; unknown reference shape"
+        )
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/references.py:181]: sqlalchemy caches for us

File: app/common/expressions/forms.py [hunk 2]
```diff
@@ -91,47 +92,46 @@ def validate(self, extra_validators=None):
 
         return super().validate(extra_validators=extra_validators)
 
-    def get_expression(self, question: Question) -> ManagedExpression:
+    def get_expression(self, subject_reference: ExpressionReference) -> ManagedExpression:
         for _managed_expression in self._managed_expressions:
             if _managed_expression.name == self.type.data:
-                return _managed_expression.build_from_form(self, question)
+                return _managed_expression.build_from_form(self, subject_reference)
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/common/expressions/forms.py:99]: to clean up later

**Review by MarcUsher — APPROVED** [review]
Pair reviewed with @samuelhwilliams & @srh-sloan and tested locally - creating new managed and custom expressions for conditions & validation all work as expected, as does editing existing expressions, expected form options appearing 🥳

**Review by MarcUsher — APPROVED** [review]
